source/index.js

```diff
@@ -23,6 +23,9 @@ const defaults = {
 		throwHttpErrors: true,
 		headers: {
 			'user-agent': `${pkg.name}/${pkg.version} (https://github.com/sindresorhus/got)`
+		},
+		hooks: {
+			beforeRequest: []
```

Comment:
I strongly disagree. I'll repeat the rationale I provided earlier in the discussion:
> If we allow the setting to be an array (without providing a custom interface), I think that might provide better mechanics for anyone wanting to apply a custom merge strategy to the options for a new instance (versus the function wrapping discussed earlier).

Requiring the use of a custom interface add no value and prohibits advanced configuration scenarios.

Reply:
> If we allow the setting to be an array (without providing a custom interface)

OK. But this doesn't change anything.

> Requiring the use of a custom interface adds no value and prohibits advanced configuration scenarios.

I disagree. What advanced configuration scenarios does it prohibit?

> I think that might provide better mechanics for anyone wanting to apply a custom merge strategy to the options for a new instance (versus the function wrapping discussed earlier).

You're right here.


What if defaults were looking like:
```js
{
	options,
	methods,
	handler: (options, hooks, next) => ...
}
```

Reply:
BTW, do we really need to specify empty hooks in the defaults?

Reply:
I prefer just a simple array too.

@szmarczak Making it just an array enables the user to construct it however they want. They can order it based on a condition.

Same reason we have:

```js
got(..., {
	headers: {foo: true}
});
```

Instead of:

```js
got.headers.add({foo: true});
```

Reply:
ok

Reply:
> I did it this way, and ensured hooks are defined as arrays during normalization

So it's needed to specify them in the defaults? If not we can just get rid of that.

Reply:
Ah, I see your point now...

I suspect that there is a performance advantage to defining the defaults.

Without the defaults, an object and `knownHookEvents.length` arrays will be created for every request, resulting in more work for the garbage collector.

Further, I think that dynamically creating the properties will introduce a new [hidden class] for every request. So, I think it is best to keep the default defined.

These are certainly micro optimizations, but I don't see a disadvantage to defining the defaults. Do you?

[hidden class]: https://draft.li/blog/2016/12/22/javascript-engines-hidden-classes/

Reply:
I don't :) But look: there are other properties that get redefined: [`options.timeout`](https://github.com/sindresorhus/got/blob/98b566444c9dd103af14281ace61fe2bd17d4f49/source/normalize-arguments.js#L179), [`options.followRedirect`](https://github.com/sindresorhus/got/blob/98b566444c9dd103af14281ace61fe2bd17d4f49/source/normalize-arguments.js#L175), [`options.retry`](https://github.com/sindresorhus/got/blob/98b566444c9dd103af14281ace61fe2bd17d4f49/source/normalize-arguments.js#L118) and more. I don't think it's a big deal.

Reply:
The array approach does come with the caveat that we can't support addition or removal of hooks from an instance since the options are run through an `Object.freeze`. A new instance would have to be created to modify hooks - they **must** be provided at instantiation.

Reply:
@brandon93s I would consider that a feature. Immutability good, it prevents a lot of weird bugs.

Do you have any use-case for when you would want to modify hooks after instantiation?

Reply:
No, nothing in particular comes to mind. I'm comfortable with how it has been implemented here. I prefer the consistent API over additional methods to maintain hooks.